src/main/java/sparksoniq/jsoniq/compiler/JsoniqExpressionTreeVisitor.java

```diff
@@ -757,10 +755,10 @@ public Void visitCastExpr(JsoniqParser.CastExprContext ctx) {
     @Override
     public Void visitUnaryExpr(JsoniqParser.UnaryExprContext ctx) {
         // TODO [EXPRVISITOR] jump from unary to postfix
-        PostFixExpression mainExpression;
+        Expression mainExpression;
         UnaryExpression node;
         this.visitSimpleMapExpr(ctx.main_expr);
-        mainExpression = (PostFixExpression) this.currentExpression;
```

Comment:
This is an extremely beneficial cleanup. Looking at the previous code I was confounded what kind of a special relationship between a PostfixExpression and Unary Expression existed and I had to check our grammar.

Reply:
Thanks!

Indeed, this was a detail from the parser, which is a "breach in encapsulation".

In the parser and the AST, the precedence of the syntactic blocks is important. In the expression tree, there is no need for this any more, because the precedence is implicit from the tree itself: every child is just an Expression, period.